feature/visited/src/main/kotlin/uk/govuk/app/visited/VisitedViewModel.kt

```diff
@@ -60,10 +73,125 @@ internal class VisitedViewModel @Inject constructor(
         )
     }
 
+    fun onRemoveVisitedItem(title: String) {
+        analyticsClient.buttonFunction(
+            text = title,
+            section = SCREEN_NAME,
+            action = REMOVE_ACTION
+        )
+    }
+
+    fun onEditClick() {
+        analyticsClient.buttonFunction(
+            text = "",
+            section = SCREEN_NAME,
+            action = EDIT_BUTTON
+        )
+    }
+
+    fun onRemoveClick() {
+        analyticsClient.buttonFunction(
+            text = "",
+            section = SCREEN_NAME,
+            action = REMOVE_BUTTON
+        )
+    }
+
+    fun onSelectAllClick() {
+        analyticsClient.buttonFunction(
+            text = "",
+            section = SCREEN_NAME,
+            action = SELECT_ALL_BUTTON
+        )
+    }
+
+    fun onDeselectAllClick() {
+        analyticsClient.buttonFunction(
+            text = "",
+            section = SCREEN_NAME,
+            action = DESELECT_ALL_BUTTON
+        )
+    }
+
+    fun onDoneClick() {
+        analyticsClient.buttonFunction(
+            text = "",
+            section = SCREEN_NAME,
+            action = DONE_BUTTON
+        )
+    }
+
     fun onVisitedItemClicked(title: String, url: String) {
         viewModelScope.launch {
             visited.visitableItemClick(title = title, url = url)
         }
         analyticsClient.visitedItemClick(text = title, url = url)
     }
+
+    fun onRemove() {
+        _uiState.value?.visited?.forEach { (_, visitedItems) ->
+            visitedItems.forEach { visitedItem ->
+                if (visitedItem.isSelected) {
+                    viewModelScope.launch {
+                        visitedLocalDataSource.remove(visitedItem.title, visitedItem.url)
+                        onRemoveVisitedItem(visitedItem.title)
+                    }
+                }
+            }
+        }
+    }
+
+    fun onSelectAll() {
+        _uiState.value = _uiState.value?.copy(
+            visited = _uiState.value?.visited?.mapValues { (_, visitedItems) ->
+                visitedItems.map { it.copy(isSelected = true) }
+            },
+            hasSelectedItems = true,
+            hasAllSelectedItems = true
+        )
+    }
+
+    fun onDeselectAll() {
+        _uiState.value = _uiState.value?.copy(
```

Comment:
We could have a private function for setting `uiState.value` as the only differences I see in onSelectAll() and onDeselectAll() is the boolean? private fun setUistate(selected: Boolean) { ... } which is called by the 2 afore mentioned functions maybe?